CASSANDRA-20389 prevent too long table names not fitting file names

k-rus opened the pull request:
The length of table names was not controlled. This is likely due to confusion between validation functions with similar names. As result creating a table with too long name led to the too long file name exception during the table creation.

This commit adds a validation of table name lengths to avoid the too long file name errors. The validation length is based on how the table name is used to create file/directory name and needs to be exact to prevent the too long file name exception, but allow all other table names, which don't lead to the too long file name exception. Since this limit is different from the existing name length limit of 48 characters, the validation of lengths is separated from the common validation function, which checks for empty name and valid characters. This separation also allows to be specific with errors in number of cases.

Note that there will be more PRs with more cases for different name length:
- Index names participate in file names together with index meta information such as index type, version and includes optional parameters, thus the length will vary on exact index created.

New constants are added for the table name length limit and for the file name length limit.
Two similar validation functions are replaced with one to avoid confusion. Name related constants are stored in the same utility class.

Table name related code are moved into methods in `TableMetadata` class, so their semantics are more clear and to allow reuse, e.g., in asserting the table name length constant.
Tests are added for the long table names and non-alphanumeric names.

Keyspace name validation function is shared between two classes and a unit test of it is added.

Other changes in this commit:
- Fix few minor suggestions for code improvements in affected files
- Add git-ignoring VS code generated files

patch by Ruslan Fomkin; reviewed by Piotr Kołaczkowski, Dmitry Konstantinov, Maxwell Guo for CASSANDRA-20389

k-rus committed:
Improve code according IDEA linter suggestions

Includes:
- Simplify passing methods as lambdas.
- Use chars instead of single char strings.
- Simplify assertion method calls.
- Remove unnecessary method throws declarations.
- Add VScode local folder to gitignore.
- Fix typo in a comment

pkolaczk commented on src/java/org/apache/cassandra/schema/TableMetadata.java:
```diff
@@ -688,7 +693,40 @@ public ClusteringComparator partitionKeyAsClusteringComparator()
+    }
+
+    /**
+     * Gets the index name from the name of the index table including dot prefix.
```
This is unclear to me. 
What is the "dot prefix"? Just the dot or also what's before the dot (that is, the table name).
Maybe you wanted to say "returns the index name prefixed by a dot or null if not an index table" ?

However, I'm not sure what's the point of dedicating a whole method for returning the index name with the leading dot. Prefixing with dot looks like something that should be done by the caller.

k-rus replied:
Good point about the confusing description. I also agree that the dot is specific to the caller, however, removing the dot will bloat the code.

pkolaczk commented on src/java/org/apache/cassandra/schema/SchemaConstants.java:
```diff
@@ -75,14 +77,54 @@ public final class SchemaConstants
+     *
+     * @param name      the name to check
+     * @param maxLength max acceptable length for the given name. For the cases when it cannot
+     *                  be limited, 0 or negative number should be supplied.
```
For the cases where it *cannot* be limited - did you mean "for the cases where it *should* not be limited" ?
Also, specifying zero doesn't look like natural value. Why not specify `Integer.MAX_VALUE` for unlimited and assert maxLength is > 0?

k-rus replied:
I mean cannot be limited. It actually should be limited, ~~but fixing the left case is outside the scope of this issue~~ but cannot with a single constant.

k-rus replied:
>Also, specifying zero doesn't look like natural value.

I don't follow what it means.



> Why not specify `Integer.MAX_VALUE` for unlimited and assert maxLength is > 0?

It don't see that it gives any advantage.

I think it's good to revisit when implementing CASSANDRA-20445

smiklosovic replied:
@k-rus 

_It don't see that it gives any advantage._

Because then you do not need to do 

`&& (maxLength <= 0 || name.length() <= maxLength);`

but just 

`&& (name.length() <= maxLength);`

because `name.length()` will be always < when `maxLength` is  `Integer.MAX_VALUE`

Also, who would ever call that method with `maxLength <= 0`, meaning `maxLength` being `0`? That does not make sense to call like that. `Integer.MAX_VALUE` is more "idiomatic" in such a way that we practically do not care how long it will be, it might be so long, up to `Integer.MAX_VALUE`, that we effectively do not care. 

Asserting that `maxLength > 0` at the beginning of this method is just enough imho. 

What you are doing is that you say "and put there 0 or negative if you don't care about the length". What is suggested is that you still stay working with the logic of "how long I want to have it at maximum" and you do not care so you put there insanely large number.

The latter case just parses better in the head because we are putting here special numbers, 0, -1 just to model something which is perfectly covered by `Integer.MAX_VALUE`. With your solution, people need to know that "we do not care equals to negative or 0". They don't need to know this. Using `Integer.MAX_VALUE` completely circumvents this.

I would also move `name.length() <= maxLength` _before_ `PATTERN_WORD_CHARS.matcher(name).matches()`.

It does not make sense to evaluate pattern when length is bigger than allowed and checking the length of a string is presumably faster than executing pattern matching. 

Also, I think that `!name.isEmpty()` is not enough. What if `name` is `     ` - meaning e.g. 5 spaces? Then `name.isEmpty()` is false but `     ` is hardly a valid file name so you would execute pattern matching unnecessarily.  You should use `!name.isBlank()` instead which is there from Java 11.

k-rus replied:
>Also, I think that !name.isEmpty() is not enough. What if name is       - meaning e.g. 5 spaces? Then name.isEmpty() is false but       is hardly a valid file name so you would execute pattern matching unnecessarily. You should use !name.isBlank() instead which is there from Java 11.

@smiklosovic 
Thank you for sharing about `isBlank()`.
Spaces and other non-alphanumeric and non-underscore are already covered by `PATTERN_WORD_CHARS.matcher(name).matches()`. From your comment it sounds that the current condition does not fail `    `, while it does. What do I miss?
(note I am not arguing about `isBlank` vs `isEmpty`, but trying to understand what I am missing)

k-rus replied:
@smiklosovic 
I disagree with your statement about didn't care about controlling the length, like in
> it might be so long, up to Integer.MAX_VALUE, that we effectively do not care.

>What you are doing is that you say "and put there 0 or negative if you don't care about the length".

>What is suggested is that you still stay working with the logic of "how long I want to have it at maximum" and you do not care so you put there insanely large number.

The point is that controlling the length is important, however, it's not always possible in this simplistic approach with a single value. E.g., it will not work for an index name, which will be clear when fixing [CASSANDRA-20445](https://issues.apache.org/jira/browse/CASSANDRA-20445). Your points for using `Integer.MAX_VALUE` gives me impression that it will not deliver the message that the controlling the length is missing here. And I don't see any case where controlling length is not important and allows to not care.

From having the discussion I think it's better to separate validation of names' content from names' length. WDYT?

k-rus replied:
> The latter case just parses better in the head because we are putting here special numbers, 0, -1 just to model something which is perfectly covered by `Integer.MAX_VALUE`. With your solution, people need to know that "we do not care equals to negative or 0". They don't need to know this. Using `Integer.MAX_VALUE` completely circumvents this.

The solution doesn't prevent from using `Integer.MAX_VALUE`, but it allows to write `isValidName(name, -1)`, which is likely to require to check what it means. However, since you missed to understand this point, I am coming to conclusion that neither `isValidName(name, -1)` nor `isValidName(name, Integer.MAX_VALUE)` is suitable.

k-rus replied:
@smiklosovic I want to let you to know that your comment produced 17 Jira and email notifications due to your comment editing. It would be helpful if you can be more mindful when you write comments and edit them. Unfortunately, I cannot do much on my side to reduce the issue.

k-rus replied:
> > Also, I think that !name.isEmpty() is not enough. What if name is       - meaning e.g. 5 spaces? Then name.isEmpty() is false but       is hardly a valid file name so you would execute pattern matching unnecessarily. You should use !name.isBlank() instead which is there from Java 11.
> 
> @smiklosovic Thank you for sharing about `isBlank()`. Spaces and other non-alphanumeric and non-underscore are already covered by `PATTERN_WORD_CHARS.matcher(name).matches()`. From your comment it sounds that the current condition does not fail ` `, while it does. What do I miss? (note I am not arguing about `isBlank` vs `isEmpty`, but trying to understand what I am missing)

Actually looking to a bigger picture, I don't think it's good idea to replace `is Empty` with `isBlank`. Some identifier names are used in file names (e.g., table and index names) and thus are restricted by `name.isEmpty` and `PATTERN_WORD_CHARS.matcher(name).matches()`, while other identifier names are not used in file names (e.g., column names), thus only restricted by `name.isEmpty`. The latter allows name like `\"   \"`, thus `name.isBlank` cannot be used. Thus to me it will be more logical to keep `isEmpty`:
- Identifier names used in file names are limited by `name.isEmpty` and `PATTERN_WORD_CHARS.matcher(name).matches()`
- Identifier names not used in file names are limited by `name.isEmpty` only

(I left out length limit from this comment)

smiklosovic replied:
@k-rus _The solution doesn't prevent from using Integer.MAX_VALUE, but it allows to write isValidName(name, -1), which is likely to require to check what it means._

Why do we even require a developer to check what it means. Why not to have just dedicated methods with wired max etc. Basically, methods tailored for what we want to express without putting there any "max sizes" ourselves.

Sorry if this question is too simplistic as I do not have a lot of context about what is going on in details.

But I could imagine have a method like "checkValidTableName(name))" and in it, we would have static "max". Similarly for other stuff.

In other words, why do we enable developers to put there their own "max"?

pkolaczk reviewed (approved):
LGTM, just a few minor code-style nitpicks ;)

k-rus commented:
@smiklosovic @pkolaczk I pushed a commit to separate name length validations from the validation for emptiness and characters. The rational is:
- Different names require different length validations and some validations will be more dynamic, e.g., table name in 4.x or index name.
- It gives opportunity to separate errors on length and on characters.
My understanding is that originally any name should fit 48 character, thus it made sense to have the length validation in the same function as the character validation. However, since the length validation was missed for some names, there can be longer names in existing databases, and thus 48 characters cannot be applied.

What do you think about the last implementation?

k-rus commented:
I updated the PR title to include the latest changes.

smiklosovic commented:
i wont be available whole next week

smiklosovic commented:
@k-rus see? I love the current approach! I will try to review this, definitely after 5th May, maybe next week but as I said availability will be limited. You are free to ask another commiter for the second review if you want. 

edit: I put some comments in.

smiklosovic commented on src/java/org/apache/cassandra/schema/TableMetadata.java:
```diff
@@ -602,11 +606,16 @@ public boolean dependsOn(Function function)
 
-        if (!isNameValid(name))
-            except("Table name must not be empty, more than %s characters long, or contain non-alphanumeric-underscore characters (got \"%s\")", SchemaConstants.NAME_LENGTH, name);
+        if (!isValidName(name))
```
@k-rus why not to have "isValidTableName" method when we have already "isValidKeyspaceName" method? If we put "keyspace" in one, we should put "table" in the other.

smiklosovic replied:
If "isValidName" is used at other places too, we could have "isValidTableName()" method in TableMetadata which would just call "isValidName" in SchemaConstants. It seems like an overkill but I think that having the naming consistent is worth of it.

k-rus replied:
Moved table name validation into a separate method. Thank you for the suggestion.

smiklosovic commented on src/java/org/apache/cassandra/schema/TableMetadata.java:
```diff
@@ -602,11 +606,16 @@ public boolean dependsOn(Function function)
-        if (!isNameValid(keyspace))
-            except("Keyspace name must not be empty, more than %s characters long, or contain non-alphanumeric-underscore characters (got \"%s\")", SchemaConstants.NAME_LENGTH, keyspace);
+        if (!isValidKeyspaceName(keyspace))
+            except("Keyspace name must not be empty, more than %s characters long, or contain non-alphanumeric-underscore characters (got \"%s\")", NAME_LENGTH, keyspace);
```
@k-rus 

what I would prefer to do is to move "except" (its guts) into "isValidKeyspaceName". If we always throw when we check, why not to move throwing to that method? Same for table version of it (together with "assert" after checking logic)

That would lead to the renaming of these methods to "static void validateKeyspaceName(String)" and "static void validateTableName(String)". This stuff in here would be reduced to two method calls. 

SchemaConstants should contain this whole helper logic. We would just call it from there here. We would make TableMetadata easier to read.

k-rus replied:
> what I would prefer to do is to move "except" (its guts) into "isValidKeyspaceName". If we always throw when we check, why not to move throwing to that method?

@smiklosovic 
Thank you for the suggestion. I investigated to throw from inside `isValidKeyspaceName`, however, the callers throw different exception classes: `ConfigurationException` and `InvalidRequestException`. One solution can be to pass the exception class, but it doesn't look to be simple in Java 11: requires to add a supplier interface. Let me know if you know a better solution.

>SchemaConstants should contain this whole helper logic. We would just call it from there here. We would make TableMetadata easier to read.

At this moment I feel strongly to keep the methods in `KeyspaceMetadata` and `TableMetadata` and not move into `SchemaConstants`:
- To me the logic is specific, e.g., table name validation is not relevant for keyspace or index name validations, while `SchemaConstants` is common utility to me. In such case, may be `TABLE_NAME_LENGTH` should be moved out from `SchemaConstants`.
- The version of this patch for 4.x will contain more complex logic for validating table name length.
- Index name length validation will be fixed by CASSANDRA-20445 and it will contain complex logic.

smiklosovic replied:
@k-rus both of them are extending RequestValidationException. Yes, supplier would be the best. Then you would throw supplier.get(); , that being RequestValidationException

smiklosovic replied:
The "shortcoming" of the  current approach is that you can not, easily, reuse / check if something is valid or not. It is "burried" in TableMetadata#validate(). If you extracted it to a helper class, it is there for everybody to use easily. 

SchemaConstants do not contain just _contstants_. It already contains validation methods etc so I feel this would be a nice fit.

smiklosovic replied:
For example, if somebody is generating names of keyspaces and tables in tests, they might run it through these validation methods in SchemaConstants to be sure it is indeed valid. I think it makes a lot of sense to have this logic extracted.

smiklosovic replied:
Well, I mean ... they could reuse it, they just would not get the exceptions thrown as in here etc ... I ll think about that more.

smiklosovic commented on src/java/org/apache/cassandra/schema/SchemaConstants.java:
```diff
@@ -41,6 +41,8 @@
 {
     public static final Pattern PATTERN_WORD_CHARS = Pattern.compile("\\w+");
+    public static final Pattern PATTERN_NON_WORD_CHAR = Pattern.compile("\\W");
+
```
@k-rus remove this redundant empty line please

k-rus replied:
Fixed

smiklosovic commented on src/java/org/apache/cassandra/schema/KeyspaceMetadata.java:
```diff
@@ -62,6 +62,12 @@ public final class KeyspaceMetadata implements SchemaElement
 {
     public static final Serializer serializer = new Serializer();
 
+    public static boolean isValidKeyspaceName(String keyspaceName)
```
@k-rus I would move this method to SchemaConstants. Check comments below as well. All validation stuff would go to SchemaConstants.

k-rus replied:
I replied in https://github.com/apache/cassandra/pull/4038#discussion_r2061246845 why I would prefer not to move. Let's see what you think about my reasoning.

smiklosovic commented on src/java/org/apache/cassandra/cql3/statements/schema/CreateIndexStatement.java:
```diff
@@ -220,6 +220,14 @@ Set<String> clientWarnings(KeyspacesDiff diff)
         return ImmutableSet.of();
     }
 
+    private void validateCustomIndexColumnName(String name)
```
to SchemaConstants

k-rus commented:
> @k-rus see? I love the current approach! I will try to review this, definitely after 5th May, maybe next week but as I said availability will be limited. You are free to ask another commiter for the second review if you want.
> 
> edit: I put some comments in.

I confirm that I understand that you will have possibility to review after next week. I think it's better to wait until you are back. Meanwhile I addressed some suggestions and replied to another.

smiklosovic commented on src/java/org/apache/cassandra/schema/TableMetadata.java:
```diff
@@ -602,11 +606,16 @@ public boolean dependsOn(Function function)
+        if (!isValidName(name))
+            except("Table name must not be empty or not contain non-alphanumeric-underscore characters (got \"%s\")", name);
+
+        if (name.length() > TABLE_NAME_LENGTH)
```
this is a little bit off, I would expect that if isValidTableName is true, then I will not be required to have yet another check where I check if name.length() > TABLE_NAME_LENGTH. That might be part of that already no?

What might happen is that a dev calls "isValidTableName(name)" and if it returns true, why would they check if length is smaller than TABLE_NAME_LENGTH. They just forget to do this. Because they just got "true" from "isValidTableName" .

In other words, by having the logic spread here like this, "isValidName" is not actually telling truth, does it?

k-rus replied:
These two checks and the assert are replaced with a single method call to `validateTableName`, which also encapsulates throwing an exception on invalid table name.

Maxwell-Guo replied:
ping @smiklosovic , do you agree with @k-rus 's fix ? changing the name from "!isValidName(name)" to "isValidCharsName"

smiklosovic commented on src/java/org/apache/cassandra/schema/TableMetadata.java:
```diff
@@ -602,11 +606,16 @@ public boolean dependsOn(Function function)
+        if (name.length() > TABLE_NAME_LENGTH)
+            except("Table name must not be more than %d characters long (got %d characters for \"%s\")", TABLE_NAME_LENGTH, name.length(), name);
+
+        assert getTableDirectoryName().length() <= FILENAME_LENGTH : String.format("Generated directory name for a table of %d characters doesn't fit the max filename legnth of %s. This unexpectedly wasn't prevented by check of the table name length, %d, to fit %d characters (got table name \"%s\" and generated directory name \"%s\"", getTableDirectoryName().length(), FILENAME_LENGTH, name.length ...
```
same with this. If I got true from "isValidTableName" there is no way I would go and check this as well.

That means what we would move three "get" methods bellow to SchemaConstants as well and accommodate their signatures to reflect the changes.

k-rus replied:
I disagree with the suggestion of moving `getTableDirectoryName` into the utility class. As it operate on a table name and used with `TableMetadata` instance.

k-rus committed:
Prevent too long table names not fitting file names

Fixes CASSANDRA-20389

The length of table names was not controlled. This is likely due to
confusion between validation methods with similar names. As result
creating tables with too long names led to the too long file name
exceptions during table creations.

This commit adds a validation of table name lengths to avoid the too
long file name errors. The validation length is based on how the table
name is used to create file/directory names, and needs to be exact to
prevent the too long file name exception, but allow all other table
names, which didn't lead to the too long file name exception. This
length limit is different from the existing name length limit of 48
characters used by common validation functions.

Thus, this commit moves out the length validation from the validation
methods into a separate length validation method, so the errors on
names are more specific. The non-length validation methods combined
into a single method, which checks for empty names and valid characters.

New constants are added for the length limits.

Table name related code are moved into methods in TableMetadata class,
so their semantics are more clear and to allow reuse, e.g., in
asserting the table name length constant.
Tests are added for the long table names and non-alphanumeric names.

Keyspace name validation function is now shared between two classes and
a unit test of it is added.

Patch by Ruslan Fomkin; reviewed by Piotr Kołaczkowski, Dmitry
Konstantinov, Maxwell Guo for CASSANDRA-20389

k-rus commented:
I found few test failures due to changes in the error message. Looking into them.

k-rus commented:
> I found few test failures due to changes in the error message. Looking into them.

The tests are in Python's DTest's repo.

netudima commented on src/java/org/apache/cassandra/schema/KeyspaceMetadata.java:
```diff
@@ -62,6 +64,32 @@ public final class KeyspaceMetadata implements SchemaElement
+        {
+            if (!SchemaConstants.isValidCharsName(keyspaceName))
+                throw ex.getConstructor(String.class)
+                        .newInstance(format("Table name must not be empty or not contain non-alphanumeric-underscore characters (got \"%s\")",
```
the method is to validate a keyspace name but the exception message it produces says: "Table name .."

netudima commented on src/java/org/apache/cassandra/schema/KeyspaceMetadata.java:
```diff
@@ -62,6 +64,32 @@ public final class KeyspaceMetadata implements SchemaElement
+     * @param keyspaceName The name of the keyspace to validate
+     * @param ex           The exception class to throw if validation fails
+     */
+    public static <T extends RequestValidationException> void validateKeyspaceName(String keyspaceName, Class<T> ex)
```
what is about using a lambda here to create Exception instance instead of reflection? If a signature for one of used exception classes change the current logic will detect it only in runtime, with a lambda it will be found during a compilation.

netudima commented on src/java/org/apache/cassandra/schema/SchemaConstants.java:
```diff
@@ -77,14 +78,32 @@ public final class SchemaConstants
 
-    public static boolean isValidName(String name)
+    /**
+     * Validates that a name contains only alphanummeric characters or underscore,
```
typo: alphanummeric

k-rus replied:
I am not sure what typo is here. Can you clarify? (there is/was a typo in this work in another file, KeyspaceMetadata)

netudima replied:
alphanummeric should be with a single m

k-rus replied:
Thank you. I confirm that it's so :) BTW you can use Suggestions in GH to make your suggestions more clear. E.g.:
```suggestion
     * Validates that a name contains only alphanumeric characters or underscore,
```

netudima commented on src/java/org/apache/cassandra/schema/SchemaConstants.java:
```diff
@@ -77,14 +78,32 @@ public final class SchemaConstants
+     * Longest acceptable table name, so it can be used in a directory
+     * name constructed with a suffix of a table id and a separator.
+     */
+    public static final int TABLE_NAME_LENGTH = FILENAME_LENGTH - 32 - 1;
```
what is about introducing an extra constant here instead of a magic number (32), like: 
TABLE_UUID_AS_HEX_LENGTH = 32
to make the logic a bit more clear?

netudima commented on src/java/org/apache/cassandra/schema/SchemaConstants.java:
```diff
@@ -77,14 +78,32 @@ public final class SchemaConstants
+     * so it can be used in file or directory names.
+     *
+     * @param name the name to check
+     * @return whether the name contains only valid characters
```
note: we also check that it is not empty

k-rus replied:
The comment at the beginning says:
>a name contains only alphanumeric characters or underscore

Isn't it enough? Or do you want to have it explicit that the name cannot be empty?

netudima replied:
Yes, I think it would be nice to add (technically an empty string contains only alphanumeric characters or underscore :-) ) but it is a nit, so up to you

k-rus replied:
While I was updating the error message, I updated this comment too.

netudima reviewed (commented):
do we have an e2e test for index creation similar to CreateTableValidationTest?

k-rus commented:
> do we have an e2e test for index creation similar to CreateTableValidationTest?

@netudima what does `e2e` means?

netudima commented:
end-to-end, a test like you have added for tables using CqlTester (CreateTableValidationTest), to ensure that the validation is actually used.

k-rus commented:
@netudima thank you for your comments. I addressed them except the empty name is invalid char.
This PR focuses on table and keyspace names, thus I don't see a need of adding index name tests. There is a separate issue for dealing with index names, [CASSANDRA-20445](https://issues.apache.org/jira/browse/CASSANDRA-20445). I suggest to address more tests there as it will be more natural.

k-rus commented:
@netudima @pkolaczk thank you for reviewing and approving this PR. Can you review the corresponding PR for DTest, https://github.com/apache/cassandra-dtest/pull/281? 

I will squash commits into 2 commit: one is minor improvement to fix linter reports on affected files, and another for the main contribution in this PR. Then I will work on porting them to affected release branches.

Maxwell-Guo commented on src/java/org/apache/cassandra/cql3/statements/schema/CreateIndexStatement.java:
```diff
@@ -60,8 +60,8 @@ public final class CreateIndexStatement extends AlterSchemaStatement
     public static final String COLUMN_DOES_NOT_EXIST = "Column '%s' doesn't exist";
-    public static final String INVALID_CUSTOM_INDEX_TARGET = "Column '%s' is longer than the permissible name length of %d characters or" +
-                                                             " contains non-alphanumeric-underscore characters";
+    public static final String INVALID_CHARS_CUSTOM_INDEX_TARGET = "Column '%s' contains non-alphanumeric-underscore characters";
```
There should be some relationship between the variable name and the variable content，But the name is INVALID_CHARS_CUSTOM_INDEX_TARGET and the content is column contains  non-alphanumeric-underscore,

k-rus replied:
I am not sure what you are trying to say. There is clear relationship as in the original constant. The error message is to be used for a customer index target as in the original variable and the error is about invalid characters. I don't see need to be more verbose in the constant and explain which chars are invalid.

Maxwell-Guo replied:
This will help the code readability. 
see this google example :
`public static final String COLUMN_DOES_NOT_EXIST = "Column '%s' doesn't exist";`

k-rus replied:
I am still not sure what needs to be improved. I also see examples, which support current naming. E.g.:
```
public static final String TOO_LONG_CUSTOM_INDEX_TARGET = "Column '%s' is longer than the permissible name length of %d characters";
```
What do you think about name `INVALID_CHARS_CUSTOM_INDEX_TARGET_COLUMN`? To me it's unnecessary verbose, since a target means a column.

Maxwell-Guo replied:
This is just a question of whether the variable naming is accurate. Although I brought it up, I don't think it's a big problem. I am ok if you don't want to change the name.

Maxwell-Guo commented on src/java/org/apache/cassandra/cql3/statements/schema/CreateIndexStatement.java:
```diff
@@ -229,8 +237,9 @@ private void validateIndexTarget(TableMetadata table, IndexMetadata.Kind kind, I
 
-        if ((kind == IndexMetadata.Kind.CUSTOM) && !SchemaConstants.isValidName(target.column.toString()))
-            throw ire(INVALID_CUSTOM_INDEX_TARGET, target.column, SchemaConstants.NAME_LENGTH);
+        // This check will be removed with CASSANDRA-20235
```
remove this line

k-rus replied:
This line is put on purpose as such check must not exist. It indicates that fixing it is left to a follow up work. Can you explain your reasoning for your comment?

Maxwell-Guo replied:
The codebase is not a notepad. You already have Jira, so why do you need this kind of meaningless record? Have you seen similar notepad messages in recent Cassandra code?

Maxwell-Guo replied:
cc @netudima @smiklosovic

k-rus replied:
> The codebase is not a notepad.

Can you explain your intention with this phrase, so I am not misinterpreting it?



> You already have Jira, so why do you need this kind of meaningless record?

To me it's not meaningless, since otherwise I will not be surprised that reviewers would complain of not doing the correct fix for this place. In general I am on the side of not having TODOs in the code and have them only in tickets. Here I find it's useful to have a comment with the reference to the ticket, and I got impression that it's preferred approach in C*. Do you want me to improve the comments?



> Have you seen similar notepad messages in recent Cassandra code?

I find 1176 comments with references to Jira's tickets in C* java code. I haven't checked how fresh are they. I also don't find any notes in the contribution style about it.

k-rus replied:
> > Have you seen similar notepad messages in recent Cassandra code?
> 
> I find 1176 comments with references to Jira's tickets in C* java code. I haven't checked how fresh are they. I also don't find any notes in the contribution style about it.

Checked one comment with TODO and it's fresh, just 9 months old:
```
// TODO: Temporary fix for CASSANDRA-20079
```

Maxwell-Guo replied:
`Can you explain your intention with this phrase, so I am not misinterpreting it?`
When I saw this comment, I felt like the author had written a memo to himself, asking him to delete it in which patch. I don't like this kind of memo.
1、it doesn’t even have a basic TODO
2、I think even the author mark this down, the author should have a basic test case to cover it . what if the author forget to remove it in CASSANDRA-20235 ? and how we can find it ? Of course, if we are lucky, there are other test cases that can cover it.

k-rus replied:
> 2、I think even the author mark this down, the author should have a basic test case to cover it . what if the author forget to remove it in [CASSANDRA-20235](https://issues.apache.org/jira/browse/CASSANDRA-20235) ? and how we can find it ? Of course, if we are lucky, there are other test cases that can cover it.

Can you clarify what test you mean? I don't think you suggest to have a test, which fails until this task is completed.

k-rus replied:
> what if the author forget to remove it in [CASSANDRA-20235](https://issues.apache.org/jira/browse/CASSANDRA-20235) ?

This is one of the reason for the comment. I don't have a better idea how to help to eventually remove the check.

k-rus replied:
I updated this comment to prefix with `TODO` as discussed in DM.

Maxwell-Guo commented on .circleci/config.yml:
I think we do not need to change the config.yml in the patch

k-rus replied:
Can you explain how the PR can be tested in CI otherwise? I well understand that this change must be removed before merge by a committer.

Maxwell-Guo replied:
You can do the test yourself, but this file will not be allowed to be submitted to the git repo, so submitting this file will bring unnecessary workload to the committer or pmc who will assist you in testing and merging the code later.

k-rus replied:
> this file will not be allowed to be submitted to the git repo, so submitting this file will bring unnecessary workload to the committer or pmc who will assist you in testing and merging the code later.

I understand that this change to the CircleCI configuration will need to be removed. Removing this change now means that the affected test will fail while working on the PR, which, to me, means more work to run CI builds.

I am not sure what is the proper approach, but previously I was advised to have such change while working on a ticket. (this change will be needed to run CI against other branches)

ekaterinadimitrova2 replied:
We normally add the bumped config for CircleCI in a separate commit with a message that contains `DO NOT COMMIT` or so, for clarity. When the patch is ready to commit, that commit is reverted. This was a common practice when most people were still using CircleCI. Nowadays it is mostly Jenkins.

ekaterinadimitrova2 replied:
>  so submitting this file will bring unnecessary workload to the committer or pmc who will assist you in testing and merging the code later.

Unfortunately, there is no other way to test in CircleCI without that additional commit. The default config in the repo is for the free tier which is not enough to run tests in CircleCI. @Maxwell-Guo , please check the circleci Readme - https://github.com/apache/cassandra/blob/trunk/.circleci/readme.md#switching-to-higher-resource-settings

Maxwell-Guo replied:
Thanks for your explanation @ekaterinadimitrova2 , What I want to express is that the submitter should provide a clearer code. If I want to test it, whether in my circle ci environment or Jenkins environment, I need to make additional modifications to the above config.yml file.

ekaterinadimitrova2 replied:
@Maxwell-Guo 

> What I want to express is that the submitter should provide a clearer code.

Yes, agreed. That is why the common practice in the project regarding CircleCI is to have the paid config submitted in a separate commit with a comment that it should not be committed. The only way to use the config in CircleCI is if you have that committed change in your branch you test. It is removed before final actual commit to Cassandra repo. 

> my circle ci environment

I am afraid this has nothing to do with any private entity environment. CircleCI is a CI adopted by the Cassandra project. Though the main one is Jenkins. How to push the paid config to be able to run Python tests is also explained as part of the readme. We were even releasing based on CircleCI runs. The main reason for that option to exist is non-committers who do not have access to the Apache Jenkins CI, and because its resources are limited. Also, it serves as a backup CI environment.
https://cassandra.apache.org/_/development/ci.html

With that said, I see that @k-rus didn't put it in a separate commit, and it can be confusing and also harder to revert it at the end. @k-rus, please ensure you add that config in a separate commit with a clear comment next time. Thanks

k-rus replied:
> With that said, I see that @k-rus didn't put it in a separate commit, and it can be confusing and also harder to revert it at the end. @k-rus, please ensure you add that config in a separate commit with a clear comment next time. Thanks

@ekaterinadimitrova2 
It's a separate commit: cef58431263362605942d771340f68b3fc2eba85 (but I missed and got a single change in another file and I will move it )

@ekaterinadimitrova2  @Maxwell-Guo Can it be good idea to put this change to CircleCi in a separate branch?

ekaterinadimitrova2 replied:
> Can it be good idea to put this change to CircleCi in a separate branch?

WDYM? 1 branch for review and one for testing? It is up to you, but I don't think it is needed. Being in a separate commit with a message something like "DO NOT COMMIT only for testing" should be enough, and it is a common practice when we use CirlceCI.

k-rus replied:
Yes, two branches. After seeing that I missed and committed another change into this commit I feel I want to use separate branch for testing to avoid such mistakes: it will worth of the overhead :)

k-rus replied:
I removed this change from this branch and put into a separate branch

Maxwell-Guo commented on src/java/org/apache/cassandra/schema/IndexMetadata.java:
```diff
@@ -111,29 +108,25 @@ public static IndexMetadata fromIndexTargets(List<IndexTarget> targets,
     public void validate(TableMetadata table)
     {
-        if (!isNameValid(name))
+        // Validating the length will be addressed by CASSANDRA-20445
```
remove this line please

k-rus replied:
This comment is added intentionally that fixing length validation for index names is left for a follow up commit. Can you provide a reasoning and way to improve here?

k-rus replied:
I updated this comment to prefix with `TODO` as discussed in DM.

Maxwell-Guo commented on src/java/org/apache/cassandra/schema/IndexMetadata.java:
```diff
@@ -117,29 +114,25 @@ public static IndexMetadata fromIndexTargets(List<IndexTarget> targets,
     public static String generateDefaultIndexName(String table)
     {
-        return PATTERN_NON_WORD_CHAR.matcher(table + "_" + "idx").replaceAll("");
+        return PATTERN_NON_WORD_CHAR.matcher(table + "_idx").replaceAll("");
```
please Keep it unchanged.

k-rus replied:
It's in a separate commit. I am not aware a better to way to do such minor fixes.

k-rus replied:
See also my comment: https://github.com/apache/cassandra/pull/4038#discussion_r2281701393

Maxwell-Guo replied:
so can you submit a new commit and just restore them to the original states?

k-rus replied:
> so can you submit a new commit and just restore them to the original states?

What do you mean? It's in a separate commit already. Do you suggest to change the order of the commits?

k-rus replied:
> > so can you submit a new commit and just restore them to the original states?
> 
> What do you mean? It's in a separate commit already. Do you suggest to change the order of the commits?

I put the commit with these not directly related changes first before the main contribution commit(s), since I expect more changes in the main contribution to address reviewers' comments. So I can just add new commits addressing comments and squash them later to the main contribution commit.

Maxwell-Guo commented on src/java/org/apache/cassandra/schema/IndexMetadata.java:
```diff
@@ -117,29 +114,25 @@ public static IndexMetadata fromIndexTargets(List<IndexTarget> targets,
-
     public static String generateDefaultIndexName(String table, ColumnIdentifier column)
     {
-        return PATTERN_NON_WORD_CHAR.matcher(table + "_" + column.toString() + "_idx").replaceAll("");
```
please Keep it unchanged.

k-rus replied:
It's in a separate commit. I am not aware a better to way to do such minor fixes.

k-rus replied:
See also my comment: https://github.com/apache/cassandra/pull/4038#discussion_r2281701393

Maxwell-Guo replied:
same above

Maxwell-Guo commented on src/java/org/apache/cassandra/schema/IndexMetadata.java:
```diff
@@ -117,29 +114,25 @@ public static IndexMetadata fromIndexTargets(List<IndexTarget> targets,
     {
         Map<String, String> newOptions = new HashMap<>(options);
         newOptions.put(IndexTarget.TARGET_OPTION_NAME, targets.stream()
-                                                              .map(target -> target.asCqlString())
```
Why modify this?

k-rus replied:
This fixes a linter complain and, imho, improves code readability. It's in a separate commit and I am not aware about a better way of fixing such minor issue.

k-rus replied:
See also my comment: https://github.com/apache/cassandra/pull/4038#discussion_r2281701393

Maxwell-Guo commented on src/java/org/apache/cassandra/schema/TableMetadata.java:
```diff
@@ -649,6 +651,17 @@ public void validate()
+        if (name.length() > TABLE_NAME_LENGTH)
+            except("Table name must not be more than %d characters long (got %d characters for \"%s\")", TABLE_NAME_LENGTH, name.length(), name);
+
+        assert getTableDirectoryName().length() <= FILENAME_LENGTH : String.format("Generated directory name for a table of %d characters doesn't fit the max filename legnth of %s. This unexpectedly wasn't prevented by check of the table name length, %d, to fit %d characters (got table name \"%s\" and generated directory name \"%s\"", getTableDirectoryName().length(), FILENAME_LENGTH, name.length ...
```
why you assert here and exception on other two cases?

k-rus replied:
> why you assert here and exception on other two cases?

I don't understand the question. Can you elaborate? 
This is the only place where length is controlled to fit the file name length, isn't it?

Maxwell-Guo replied:
My question here is , you throw exception for line 656 and 659 but doing assert validation at line 662 , why not throw an exception too?

k-rus replied:
This is already covered by the exception on line 659. This assert is to prevent introducing a bug on line 659. Maybe I am overcautious here and the assert can be removed. I was hoping that the assert will give more clarity than confusion. As you see it differently what will be preferred?

Maxwell-Guo commented on src/java/org/apache/cassandra/schema/TableMetadata.java:
```diff
@@ -813,7 +860,7 @@ public TableMetadata withUpdatedUserType(UserType udt)
     protected void except(String format, Object... args)
     {
-        throw new ConfigurationException(keyspace + "." + name + ": " + format(format, args));
+        throw new ConfigurationException(keyspace + '.' + name + ": " + format(format, args));
```
why you change to this ?

k-rus replied:
Are you asking why I think a char constant is better than a string with a single char? Or why is this change done in this PR?
For the latter I tried to briefly explain in the PR and in the commit message that it's a fix of a linter warning in a related file. I didn't go into details of the motivation, which is:

I expect that my contribution needs to be of good quality and using an IDEA linter is helpful tool to improve code. However, many warnings appear on untouched lines, which makes it harder to focus on the warnings related to my own changes. In general I believe it's good to improve code quality eventually. I understand that it's preferable to do it in separate PRs, but the overhead of doing a PR for such minor fixes is huge.
As a tradeoff, having a separate commit with these fixes within the same PR could be a good solution. Otherwise, I doubt these issues will be addressed.

Maxwell-Guo commented on src/java/org/apache/cassandra/schema/TableMetadata.java:
```diff
@@ -1799,7 +1846,7 @@ public String primaryKeyAsCQLLiteral(ByteBuffer partitionKey, Clustering<?> clus
         }
 
-        return i == 1 ? literals[0] : "(" + String.join(", ", literals) + ")";
+        return i == 1 ? literals[0] : '(' + String.join(", ", literals) + ')';
```
why you change to this ? "(" -> '('

k-rus replied:
See my comment: https://github.com/apache/cassandra/pull/4038#discussion_r2281701393

Maxwell-Guo commented on .gitignore:
```diff
@@ -57,6 +57,9 @@ ide/nbproject/private
 nb-configuration.xml
 nbactions.xml
 
+#VS code
```
i would like to create a new jira on change the gitingore , as this code has nothing to do with the pr.
But I'd also like to hear other people's opinions here.

k-rus replied:
Sounds that it's controversial to use VS code and exclude its local files, while it's acceptable to use IntelliJ, Eclipse, NetBeans. Note it's already part of [cassandra-dtest](https://github.com/apache/cassandra-dtest), i.e., it was accepted by the community.
This change is in a different commit from the main contribution and without this change it will be easy to commit files from `.vscode` folder, which will be garbage to the project.
I can create a separate Jira ticket and mention in the commit, but it will be overkill for everyone to manage additional ticket. Also separate PR is unfortunately an overkill for such minor change with current overhead.

It would be nice that such issues were already fixed and I and other new contributors can just focus on the main work of a ticket. Other approach is a quick process to deliver simple fixes. Currently I am horrified how much time I will need to invest and how long time it will take to deliver such couple of lines change.

Maxwell-Guo replied:
I think we welcome any development tools, even vim, emacs.

If there is a dispute, another committer will need to intervene. @netudima @ekaterinadimitrova2 @beobal

netudima replied:
While technically I agree that this is a non-related change I also think that a separate ticket/review and merge across branches for this very simple non-code change looks like an overkill here. 
My personal opinion: we need a balance between strict rules and usability to make the contribution process comfortable enough, so I personally prefer to focus on functional/performance and maintenance issues and trying to not dig too much into style/taste comments..

Maxwell-Guo replied:
Then I have no opinion

ekaterinadimitrova2 replied:
> Sounds that it's controversial to use VS code and exclude its local files, while it's acceptable to use IntelliJ, Eclipse, NetBeans. Note it's already part of [cassandra-dtest](https://github.com/apache/cassandra-dtest), i.e., it was accepted by the community.
> This change is in a different commit from the main contribution and without this change it will be easy to commit files from .vscode folder, which will be garbage to the project.

While in general we opt out of adding outside things to a patch, this one makes sense to me. We just stated clearly in the ticket that there is an additional change and added it in a separate commit for visibility. Thanks

k-rus replied:
> While in general we opt out of adding outside things to a patch, this one makes sense to me. We just stated clearly in the ticket that there is an additional change and added it in a separate commit for visibility.

So in addition to mentioning in the PR I should mention this in the Jira's ticket. Correct me if I misunderstood.

Similarly it's not enough that the gitignore is in a commit with changes not directly related to the contribution (https://github.com/apache/cassandra/pull/4038/commits/68b6eb1d3c070d4ed3818f4a9726b97e4315c086), but in a separate commit, so there will be three commits in the PR (VScode in gitignore, linter warning fixes in affected files, and the main contribution for the ticket). Also correct me if I misunderstood.

k-rus replied:
> > While in general we opt out of adding outside things to a patch, this one makes sense to me. We just stated clearly in the ticket that there is an additional change and added it in a separate commit for visibility.
> 
> So in addition to mentioning in the PR I should mention this in the Jira's ticket. Correct me if I misunderstood.
> 
> Similarly it's not enough that the gitignore is in a commit with changes not directly related to the contribution ([68b6eb1](https://github.com/apache/cassandra/commit/68b6eb1d3c070d4ed3818f4a9726b97e4315c086)), but in a separate commit, so there will be three commits in the PR (VScode in gitignore, linter warning fixes in affected files, and the main contribution for the ticket). Also correct me if I misunderstood.

Done

Maxwell-Guo commented:
@beobal  would you mind take a look ?

Maxwell-Guo commented:
I am + 1 if all my comments are addressed and ci is green . But i think it is better if  @beobal or @ekaterinadimitrova2 take another look.

k-rus commented:
> I am + 1 if all my comments are addressed and ci is green

Thank you for the review and finding my implementation for the ticket suitable as all the comments are not related to the main contribution. I hope it will be possible to find a reasonable way to deliver those changes as I find important to deliver them.



> But i think it is better if @beobal or @ekaterinadimitrova2 take another look.

As new reviewers might disagree with the current main implementation for the ticket I will wait with running an extensive CI.

k-rus commented:
> > I am + 1 if all my comments are addressed and ci is green
> 
> Thank you for the review and finding my implementation for the ticket suitable as all the comments are not related to the main contribution. I hope it will be possible to find a reasonable way to deliver those changes as I find important to deliver them.

I can think that adding VScode to ignore can be done in a separate ticket and PR, but it will be blocker for this PR until it's committed to trunk and release branches. Other changes do not make much sense outside this PR and will be difficult to deliver, so I still think that the separate commit is a good tradeoff.
@Maxwell-Guo what do you think?

ekaterinadimitrova2 commented:
> I am + 1 if all my comments are addressed and ci is green . But i think it is better if @beobal or @ekaterinadimitrova2 take another look.

I don't mind reviewing, but we already have three committers doing a review here. Plus @pkolaczk, who is also a long-time contributor. Is there anything specific you are worried about and you want to discuss?

Maxwell-Guo reviewed (approved):
+1 on green ci

k-rus committed:
Add VS code local folder to gitignore

k-rus commented:
I ported this patch to 5.0, see #4325. It failed a test on the error message for keyspace of `null` when called for create table, so I adjusted `validateKeyspaceName` to not change the exception message.
This situation cannot happen on `trunk`, still I ported the change back in 0f7b756f4b9c6d59e22cf78b3d9d8ba20b2e890a
However, it cannot be tested.

@pkolaczk @netudima @Maxwell-Guo ~~Do you want to review the commit and re-review the PR?~~

k-rus commented:
> I ported this patch to 5.0, see #4325. It failed a test on the error message for keyspace of `null` when called for create table, so I adjusted `validateKeyspaceName` to not change the exception message. This situation cannot happen on `trunk`, still I ported the change back in [0f7b756](https://github.com/apache/cassandra/commit/0f7b756f4b9c6d59e22cf78b3d9d8ba20b2e890a) However, it cannot be tested.
> 
> @pkolaczk @netudima @Maxwell-Guo Do you want to review the commit and re-review the PR?

Actually I found another bug in this implementation. `ire` doesn't throw, so I need to add throw.

k-rus commented:
Now I fixed that the keyspace validation throws the same exceptions as before this PR. See c9321b57c4fe078561cf9bcbed15dd2612277c6d

Unfortunately I don't find a good way to add tests on the error message in such case, since it's about internal calls as normally keyspaces are not created in the same API call for creating tables.

Also some formatting doesn't look great. I have double checked that this is in accordance with `generate-idea-files`. Since I am not fun of manual formatting I use auto-format on my changes and thus it doesn't work to format manually.

The result of running CI doesn't show any related failures: http://ci-cassandra.infra.datastax.com/job/cassandra/25/testReport/

@pkolaczk @netudima @Maxwell-Guo do you want to review the new commit? c9321b57c4fe078561cf9bcbed15dd2612277c6d

(the commit will be squashed after I am sure that the work is accepted)

k-rus commented:
While porting this patch to 4.1 I noticed one more place where I missed to throw exception on keyspace validation error. As result I refactored this PR to throw always from inside the keyspace validation and not rely on passed exception throwers from callers. See this commit: f53e6cfc798eb038bdf71539e9a0b945b876baf5
CI build can be found at http://ci-cassandra.infra.datastax.com/job/cassandra/29/testReport/ (no related errors to my understanding)
Let me know if you have any questions about it.

I plan to port this commit to 5.0 and 4.1 work, and will squash commits here.

cc @pkolaczk @netudima @Maxwell-Guo

pkolaczk reviewed (approved):
Makes sense. 
Yes, moving throw to inside validate is much cleaner and less fragile than before.

k-rus commented:
merged
